restler/engine/core/request_utilities.py

```diff
@@ -50,56 +54,71 @@ def str_to_hex_def(val_str):
-    @type  cmd: Str
-
-    @return: The result of the command
-    @rtype : Str
 
-    """
+def execute_token_refresh(token_dict):
     global latest_token_value, latest_shadow_token_value
-    _RAW_LOGGING(f"Will refresh token: {cmd}")
-
-    MAX_RETRIES = 5
-    RETRY_SLEEP_TIME_SEC = 2
     ERROR_VAL_STR = 'ERROR\r\n'
-    retry_count = 0
-    while retry_count < MAX_RETRIES:
-        try:
-            if sys.platform.startswith('win'):
-                cmd_result = subprocess.getoutput(str(cmd).split(' '))
-            else:
-                cmd_result = subprocess.getoutput([cmd])
+    result = None
+    token_auth_method = token_dict["token_auth_method"]
 
-            _RAW_LOGGING(f"New value: {cmd_result}")
-            _, latest_token_value, latest_shadow_token_value = parse_authentication_tokens(cmd_result)
-            _RAW_LOGGING(f"Successfully obtained the latest token")
+    retry_handler = RetryHandler()
+
+    while retry_handler.can_retry():
+        try:
+            if token_auth_method == "location":
+                result = execute_location_token_refresh(token_dict["location"])
+            elif token_auth_method == "cmd":
+                result = execute_token_refresh_cmd(token_dict["token_refresh_cmd"])
+            elif token_auth_method == "module":
+                result = execute_token_refresh_module(token_dict["token_module_file"], token_dict["token_module_method"], token_dict["token_module_data"])
             break
-        except subprocess.CalledProcessError:
-            error_str = f"Authentication failed when refreshing token:\n\nCommand that failed: \n{cmd}"
-            print(f'\n{error_str}')
-            latest_token_value = ERROR_VAL_STR
-            latest_shadow_token_value = ERROR_VAL_STR
-            _RAW_LOGGING(error_str)
-            retry_count = retry_count + 1
-            time.sleep(RETRY_SLEEP_TIME_SEC)
         except EmptyTokenException:
             error_str = "Error: Authentication token was empty."
             print(error_str)
             _RAW_LOGGING(error_str)
             sys.exit(-1)
         except Exception as error:
-            error_str = f"Exception refreshing token with cmd {cmd}.  Error: {error}"
-            print(error_str)
+            error_str = f"Authentication failed when refreshing token:\n\nUsing Token authentication method: \n{token_auth_method} \n with error {error}"
+            print(f'\n{error_str}')
+            latest_token_value = ERROR_VAL_STR
+            latest_shadow_token_value = ERROR_VAL_STR
             _RAW_LOGGING(error_str)
-            sys.exit(-1)
-    else:
-        _RAW_LOGGING(f"\nMaximum number of retries ({MAX_RETRIES}) exceeded. Exiting program.")
-        sys.exit(-1)
+            retry_handler.wait_for_next_retry()
+    _, latest_token_value, latest_shadow_token_value = parse_authentication_tokens(result)
+
+def execute_location_token_refresh(location):
+    f = open(location,"r")
```

Comment:
better naming. what is this location? previously used mechanism where we write token to a place and read from it?

Reply:
yeah its the location to a token stored in a file, do you have any suggestions for the naming?